File: Android/PingHTTPTests.java
import android.support.v4.util.Pair;

import com.ticketmaster.presencelibrary.PingHTTP.Get;
import com.ticketmaster.presencelibrary.PingHTTP.Post;

import org.junit.Test;

import java.util.ArrayList;

import static org.junit.Assert.assertTrue;
import static org.junit.Assert.assertFalse;

/**
 * Created by Michael.Estes on 7/19/16.
 */

public class PingHTTPTests {

    @Test
    public void getValidTest() {
        Get get = new Get("url", "endpoint");
        assertTrue(get.isValid());
    }

    @Test
    public void getInvalidTest() {
        Get get = new Get("", "");
        assertFalse(get.isValid());
    }

    @Test
    public void getInvalidTestTwo() {
        Get get = new Get("url", "");
        assertFalse(get.isValid());
    }

    @Test
    public void postValidTest() {
        Post post = new Post("url", "endpoint", "body");
        assertTrue(post.isValid());
    }

    @Test
    public void postValidTestTwo() {
        ArrayList<Pair<String, String>> params = new ArrayList<>();
        params.add(new Pair<>("testParamOne", "testParamOne"));
        params.add(new Pair<>("testParamTwo", "testParamTwo"));
        Post post = new Post("url", "endpoint", "body", params, 0);
        assertTrue(post.isValid());
    }

    @Test
    public void postInvalidTest() {
        Post post = new Post("", "", "");
        assertFalse(post.isValid());
    }

    @Test
    public void postInvalidTestTwo() {
        Post post = new Post("url", "", "");
        assertFalse(post.isValid());
    }

    @Test
    public void postInvalidTestThree() {
        Post post = new Post("url", "endpoint", "");
        assertFalse(post.isValid());
    }


    @Test
    public void reqInvalidTest() {
        Get get = new Get("", "");
        assertFalse(get.req() == new Pair<>("Error", -1));
    }

    @Test
    public void formatParamsValidTest() {
        final String expectedFormattedParams = "?name=value&name2=value2";
        ArrayList<Pair<String, String>> params = new ArrayList<>();
        params.add(new Pair<>("name", "value"));
        params.add(new Pair<>("name2", "value2"));
        Get get = new Get("url", "endpoint", params, 0);
        assertTrue(get.formatParams().equals(expectedFormattedParams));
    }

    @Test
    public void formatParamsInvalidTest() {
        ArrayList<Pair<String, String>> params = new ArrayList<>();
        Get get = new Get("url", "endpoint", params, 0);
        assertTrue(get.formatParams().isEmpty());
    }

    @Test
    public void connectionGetHTTPValidTest() {
        ArrayList<Pair<String, String>> params = new ArrayList<>();
        params.add(new Pair<>("testParamOne", "testParamOne"));
        params.add(new Pair<>("testParamTwo", "testParamTwo"));

        ArrayList<Pair<String, String>> properties = new ArrayList<>();
        properties.add(new Pair<>("Test-Authorization", "testAuth"));
        properties.add(new Pair<>("test-api-key", "testApiKey"));

        Get get = new Get("http://httpbin.org", "/get", properties, params);
        Pair<String, Integer> response = get.req();

        assertTrue(response.second == 200);
    }

    @Test
    public void connectionGetHTTPNoParamsValidTest() {
        ArrayList<Pair<String, String>> properties = new ArrayList<>();
        properties.add(new Pair<>("Test-Authorization", "testAuth"));
        properties.add(new Pair<>("test-api-key", "testApiKey"));

        Get get = new Get("http://httpbin.org", "/get", properties);
        Pair<String, Integer> response = get.req();

        assertTrue(response.second == 200);
    }

    @Test
    public void connectionGetHTTPNoPropertiesValidTest() {
        Get get = new Get("http://httpbin.org", "/get", null);
        Pair<String, Integer> response = get.req();

        assertTrue(response.second == 200);
    }

    @Test
    public void connectionGetHTTPNoPropertiesValidTestTwo() {
        ArrayList<Pair<String, String>> properties = new ArrayList<>();
        Get get = new Get("http://httpbin.org", "/get", properties);
        Pair<String, Integer> response = get.req();

        assertTrue(response.second == 200);
    }

    @Test
    public void connectionGetHTTPSValidTest() {
        ArrayList<Pair<String, String>> params = new ArrayList<>();
        params.add(new Pair<>("testParamOne", "testParamOne"));
        params.add(new Pair<>("testParamTwo", "testParamTwo"));

        ArrayList<Pair<String, String>> properties = new ArrayList<>();
        properties.add(new Pair<>("Test-Authorization", "testAuth"));
        properties.add(new Pair<>("test-api-key", "testApiKey"));

        Get get = new Get("https://httpbin.org", "/get", properties, params);
        Pair<String, Integer> response = get.req();

        assertTrue(response.second == 200);
    }

    @Test
    public void connectionGetInvalidTest() {
        ArrayList<Pair<String, String>> params = new ArrayList<>();
        params.add(new Pair<>("testParamOne", "testParamOne"));
        params.add(new Pair<>("testParamTwo", "testParamTwo"));

        ArrayList<Pair<String, String>> properties = new ArrayList<>();
        properties.add(new Pair<>("Test-Authorization", "testAuth"));
        properties.add(new Pair<>("test-api-key", "testApiKey"));

        Get get = new Get("badurl", "/get", properties, params);
        Pair<String, Integer> response = get.req();

        assertTrue(response.second == -1);
    }

    @Test
    public void connectionPostHTTPValidTest() {
        ArrayList<Pair<String, String>> params = new ArrayList<>();
        params.add(new Pair<>("testParamOne", "testParamOne"));
        params.add(new Pair<>("testParamTwo", "testParamTwo"));

        ArrayList<Pair<String, String>> properties = new ArrayList<>();
        properties.add(new Pair<>("Test-Authorization", "testAuth"));
        properties.add(new Pair<>("test-api-key", "testApiKey"));

        String body = "Test body";

        Post post = new Post("http://httpbin.org", "/post", body ,properties, params);
        Pair<String, Integer> response = post.req();

        assertTrue(response.second == 200);
    }

    @Test
    public void connectionPostHTTPSValidTest() {
        ArrayList<Pair<String, String>> params = new ArrayList<>();
        params.add(new Pair<>("testParamOne", "testParamOne"));
        params.add(new Pair<>("testParamTwo", "testParamTwo"));

        ArrayList<Pair<String, String>> properties = new ArrayList<>();
        properties.add(new Pair<>("Test-Authorization", "testAuth"));
        properties.add(new Pair<>("test-api-key", "testApiKey"));

        String body = "Test body";

        Post post = new Post("https://httpbin.org", "/post", body ,properties, params);
        Pair<String, Integer> response = post.req();

        assertTrue(response.second == 200);
    }

    @Test
    public void connectionPostInvalidTest() {
        ArrayList<Pair<String, String>> params = new ArrayList<>();
        params.add(new Pair<>("testParamOne", "testParamOne"));
        params.add(new Pair<>("testParamTwo", "testParamTwo"));

        ArrayList<Pair<String, String>> properties = new ArrayList<>();
        properties.add(new Pair<>("Test-Authorization", "testAuth"));
        properties.add(new Pair<>("test-api-key", "testApiKey"));

        String badBody = "";

        Post post = new Post("badurl", "/get", badBody, properties, params);
        Pair<String, Integer> response = post.req();

        assertTrue(response.second == -1);
    }

    @Test
    public void connectionPostInvalidTestTwo() {
        ArrayList<Pair<String, String>> params = new ArrayList<>();
        params.add(new Pair<>("testParamOne", "testParamOne"));
        params.add(new Pair<>("testParamTwo", "testParamTwo"));

        ArrayList<Pair<String, String>> properties = new ArrayList<>();
        properties.add(new Pair<>("Test-Authorization", "testAuth"));
        properties.add(new Pair<>("test-api-key", "testApiKey"));

        String body = "valid body";

        Post post = new Post("badurl", "/get", body, properties, params);
        Pair<String, Integer> response = post.req();

        assertTrue(response.second == -1);
    }
}

File: Swift/PingHTTPTests.swift
//
//  PingHTTPTests.swift
//  PingHTTP
//
//  Created by Michael Estes on 7/15/16.
//

import Foundation
import XCTest

class PingHTTPTests: XCTestCase {
    
    let genesisGet: GenesisGet = GenesisGet()
    
    override func setUp() {
        super.setUp()
    }
    
    override func tearDown() {
        super.tearDown()
    }
    
    func testGetRequestValid() {
        let testGet: Get = Get(url: "url", endpoint: "endpoint")
        XCTAssertTrue(testGet.isValid())
    }
    
    func testGetRequestInvalid() {
        let testGet: Get = Get(url: "", endpoint: "")
        XCTAssertFalse(testGet.isValid())
    }
    
    func testPostRequestValid() {
        let testPost: Post = Post(url: "url", endpoint: "endpoint", body: "body")
        XCTAssertTrue(testPost.isValid())
    }
    
    func testPostRequestInvalid() {
        let testPost: Post = Post(url: "", endpoint: "", body: "")
        XCTAssertFalse(testPost.isValid())
    }
    
    func testFormatParamsValid() {
        let expectedFormattedParams: String = "?name=value&name2=value2"
        let params: [(String, String)] = [("name", "value"), ("name2", "value2")]
        let testGet: Get = Get(url: "url", endpoint: "endpoint", params: params)
        XCTAssertTrue(testGet.formatParams() == expectedFormattedParams)
    }
    
    func testFormatParamsInvalid() {
        let params: [(String, String)] = Array()
        let testGet: Get = Get(url: "url", endpoint: "endpoint", params: params)
        XCTAssertTrue(testGet.formatParams().isEmpty)
    }
    
    func testConnectionGetValid() {
        let asyncExpectation = expectationWithDescription("longRunningFunction")
        var testStatus: Int = -1
        
        let properties: [(String, String)] = [("Test-Authorization", "testAuth"), ("test-api-key", "testApiKey")]
        let params: [(String, String)] = [("testParamOne", "testParamOne"), ("testParamTwo", "testParamTwo")]
        let get: Get = Get(url: "https://httpbin.org", endpoint: "/get", properties: properties, params: params)
        get.req { (response, status) in
            testStatus = status
            asyncExpectation.fulfill()
        }
        self.waitForExpectationsWithTimeout(5) { error in
            XCTAssertTrue(testStatus == 200)
        }
    }
    
    func testConnectionGetInvalid() {
        let asyncExpectation = expectationWithDescription("longRunningFunction")
        var testStatus: Int = -1
        
        let properties: [(String, String)] = [("Test-Authorization", "testAuth"), ("test-api-key", "testApiKey")]
        let params: [(String, String)] = [("testParamOne", "testParamOne"), ("testParamTwo", "testParamTwo")]
        let get: Get = Get(url: "badurl", endpoint: "/get", properties: properties, params: params)
        get.req { (response, status) in
            testStatus = status
            asyncExpectation.fulfill()
        }
        self.waitForExpectationsWithTimeout(5) { error in
            XCTAssertTrue(testStatus == -1)
        }
    }
}
